designcompose/src/main/java/com/android/designcompose/squoosh/SquooshAnimate.kt

```diff
@@ -132,173 +145,240 @@ internal class SquooshAnimationControl(
+    if (requestedAnimations.isEmpty()) return from.cloneSelfAndChildren(parent)
+
+    // Does this match a thing that we need to animate?
+    var requestedAnim = requestedAnimations[from.view.id]
+    if (requestedAnim == null) requestedAnim = requestedAnimations[from.unresolvedNodeId]
+    val (cloned: SquooshResolvedNode, alreadyBuiltChildren) =
+        if (requestedAnim != null && requestedAnim.animationControl == null) {
+            // Can we find it in the "to" tree?
+            val toNode = findNode(to, requestedAnim.toNodeId)
+            if (toNode != null) {
+                val animations: ArrayList<SquooshAnimatedItem> = ArrayList()
+                val mergedClonedNode = mergeRecursive(from, toNode, parent, animations)
+                if (animations.isNotEmpty())
+                    requestedAnim.animationControl =
+                        SquooshAnimationControl(mergedClonedNode, animations)
+
+                Pair(mergedClonedNode, true)
+            } else {
+                // An animation was requested, but no destination node could be found. We'll end up
+                // leaving the animationControl field blank, which will result in an error log later
+                // on.
+                Pair(from.cloneSelf(parent), false)
+            }
+        } else {
+            // Clone the target node, then look at its siblings, then at its child.
+            Pair(from.cloneSelf(parent), false)
+        }
+
+    // First recurse along siblings
+    val nextFrom = from.nextSibling
+    if (nextFrom != null) {
+        val nextCloned = createMergedAnimationTree(nextFrom, to, requestedAnimations, parent)
+        // Insert at the end; does this reorder? Hope not
+        var c: SquooshResolvedNode? = cloned
+        while (c?.nextSibling != null) c = c.nextSibling
+        c!!.nextSibling = nextCloned
     }
-    return null
+
+    // Then recurse along children, if we didn't already consider them by finding a match.
+    if (!alreadyBuiltChildren) {
+        val childFrom = from.firstChild
+        if (childFrom != null) {
+            val child = createMergedAnimationTree(childFrom, to, requestedAnimations, cloned)
+            cloned.firstChild = child
+        }
+    }
+
+    return cloned
 }
 
-// Find nodes where we want to have animation; this is typically from a component CHANGE_TO
-private fun findNode(n: SquooshResolvedNode, id: String): SquooshResolvedNode? {
-    if (n.view.id == id || n.unresolvedNodeId == id) return n
-    var child = n.firstChild
+private fun SquooshResolvedNode.cloneSelf(parent: SquooshResolvedNode?): SquooshResolvedNode =
+    SquooshResolvedNode(
+        view = this.view,
+        style = this.style,
+        layoutId = this.layoutId,
+        textInfo = this.textInfo,
+        unresolvedNodeId = this.unresolvedNodeId,
+        layoutNode = this,
+        parent = parent,
+        computedLayout = this.computedLayout,
+        needsChildRender = this.needsChildRender
```

Comment:
Do you want to set `firstChild` and `nextSibling` to `null`?

Reply:
No, I think they're always initialized to `null` anyway